tests/kernel/wave/wave_attention_test.py

```diff
@@ -561,3 +561,174 @@ def repeat(
+            imm_f16 = tkw.cast(e_delta, tkl.f8e4m3fnuz)
+            v_reg = tkw.read(v, elements_per_thread=LOAD_ELEMS_PER_THREAD)
+            v_reg = tkw.cast(v_reg, tkl.f8e4m3fnuz)
+            new_acc = acc * e_delta_max
+            acc = tkw.mma(v_reg, imm_f16, new_acc, mfma_variant[1])
+            return m_j, d_j, acc
+
+        # repeat represents the results of the loop
+        res_max, res_sum, res_mm = repeat
+        res = res_mm / res_sum
+        tkw.write(res, c, mapping=mapping, elements_per_thread=STORE_ELEMS_PER_THREAD)
+
+    hyperparams = {
+        ADDRESS_SPACE: SHARED_ADDRESS_SPACE,
+        LOAD_ELEMS_PER_THREAD: get_mfma_load_elems_per_thread(mfma_variant[0]),
+        STORE_ELEMS_PER_THREAD: get_mfma_store_elems_per_thread(mfma_variant[1]),
+        BLOCK_B: 1,
+        BLOCK_M: 128,
+        BLOCK_N: 64,
+        BLOCK_K2: 64,
+        B: shape[0],
+        M: shape[1],
+        N: shape[2],
+        K1: shape[3],
+        K2: shape[4],
+        READ_SHARED_DELAY: 1,
+        WRITE_SHARED_DELAY: 1,
+        READ_GLOBAL_DELAY: 2,
+        WRITE_GLOBAL_DELAY: 2,
+        MMA_DELAY: 1,
+        VALU_DELAY: 1,
+        SHUFFLE_DELAY: 1,
+        SHARED_MEMORY_UNITS: 4,
+        GLOBAL_MEMORY_UNITS: 4,
+        MMA_UNITS: 4,
+        VALU_UNITS: 2,
+        SHUFFLE_UNITS: 2,
+    }
+    config = get_default_run_config()
+    if run_bench:
+        config["benchmark_batch_size"] = 10
+        config["benchmark_repetitions"] = 3
+    if dump_perf is not None:
+        perf_filename = request.node.name + ".json"
+        config["benchmark_results_file"] = os.path.join(
+            dump_perf, "tk_" + perf_filename
+        )
+    with tk.gen.TestLaunchContext(
+        hyperparams,
+        canonicalize=True,
+        run=True,
+        run_bench=run_bench,
+        run_config=config,
+        schedule=enable_scheduling,
+        use_scheduling_barriers=enable_scheduling_barriers,
+    ):
+        torch.manual_seed(0)
+        q = device_randn(shape[0], shape[1], shape[3], dtype=torch.float16)
+        k = device_randn(shape[0], shape[4], shape[3], dtype=torch.float16)
+        v = device_randn(shape[0], shape[4], shape[2], dtype=torch.float16)
+        output = device_zeros(shape[0], shape[1], shape[2], dtype=torch.float32)
+        log2e = 1.44269504089
+        dk_sqrt = math.sqrt(1.0 / shape[3])
+        # TODO: Add scaling of QK as part of kernel.
+        # TODO: Add variant of non-transposed V attention kernel.
+        mb = base_attention(q * dk_sqrt * log2e, k, v.permute([0, 2, 1]), output)
+        torch_ref = torch.nn.functional.scaled_dot_product_attention(
+            q, k, v, attn_mask=None
+        )
+        if test_dump_generated_mlir:
+            filename = f"wave_attention_{'x'.join(map(str, shape))}.mlir"
+            with open(filename, "w") as f:
+                f.write(mb.module_op.get_asm())
+        rmse = torch.sqrt(torch.mean(torch.square(output - torch_ref)))
```

Comment:
Can you use assert_close with appropriate atol and rtol here instead of manually computing this?

Reply:
yeah, but the FP8 range/accuracy is quite limited so, not sure if atol/rtol becomes meaningful here. For example we have a case where the rtol is just ridiculously large:
```
Mismatched elements: 2613487 / 2621440 (99.7%)
Greatest absolute difference: 0.04659026861190796 at index (26, 785, 31) (up to 1e-05 allowed)
Greatest relative difference: 1510450.625 at index (2, 593, 37) (up to 1.3e-06 allowed)
(Pdb) torch_ref[2, 593, 37]
tensor(9.8498e-09, device='cuda:0')
(Pdb) output[2, 593, 37]
tensor(0.0149, device='cuda:0')
```

Reply:
Comparing the values of these tensors it qualitatively looks quite good, especially without any rescaling/adjusting of the range.
```
tensor([[ 0.0500, -0.0215,  0.0078,  ..., -0.0116, -0.0850,  0.0386],
        [ 0.0232,  0.0168,  0.0137,  ...,  0.1215,  0.0033,  0.0275],
        [-0.0044, -0.0127,  0.0993,  ...,  0.0158,  0.0504,  0.0125],
        ...,
        [ 0.0696, -0.0049, -0.0455,  ...,  0.0299, -0.0396,  0.0955],
        [-0.0473,  0.1092, -0.0237,  ...,  0.0194, -0.0045,  0.0034],
        [-0.0104,  0.0096,  0.0394,  ...,  0.0266, -0.0525,  0.0692]],
       device='cuda:0')
(Pdb) torch_ref[0,:]
tensor([[ 0.0467, -0.0231,  0.0064,  ..., -0.0106, -0.0863,  0.0408],
        [ 0.0245,  0.0132,  0.0127,  ...,  0.1255,  0.0051,  0.0280],
        [-0.0083, -0.0181,  0.1016,  ...,  0.0180,  0.0563,  0.0145],
        ...,
        [ 0.0675, -0.0112, -0.0434,  ...,  0.0295, -0.0368,  0.0981],
        [-0.0471,  0.1066, -0.0230,  ...,  0.0206, -0.0002, -0.0072],
        [-0.0089,  0.0065,  0.0376,  ...,  0.0283, -0.0549,  0.0682]],
        ```

Reply:
Sounds good, let's merge for now and iterate